sdks/java/core/src/main/java/org/apache/beam/sdk/coders/AvroCoder.java

```diff
@@ -206,6 +206,16 @@ public Schema get() {
   private final EmptyOnDeserializationThreadLocal<DatumWriter<T>> writer;
   private final EmptyOnDeserializationThreadLocal<DatumReader<T>> reader;
 
+  // Lazily re-instantiated after deserialization
+  @Nullable private transient ReflectData reflectData;
+
+  private ReflectData getReflectData() {
+    if (reflectData == null) {
```

Comment:
It actually is thread safe, because it is idempotent.

There could be a (probably irrelevant) performance cost trading always synchronizing for extremely rare redundant initialization.

But it does seem like a memoizing `Supplier` would be express the intent nicely, so I've made the change. Overall with the boilerplate I don't know if it was better, but I am happy either way.